apps/src/calc/expressionNode.js

```diff
@@ -87,89 +100,92 @@ ExpressionNode.prototype.clone = function () {
-      gloablMapping[this.value_]);
-    if (functionDef === undefined) {
-      throw new Error('No mapping for function during evaluation');
-    }
+      if (!functionDef.variables || !functionDef.expression) {
+        throw new Error('Bad mapping for: ' + this.value_);
+      }
+      if (functionDef.variables.length !== this.children_.length) {
+        throw new Error('Bad mapping for: ' + this.value_);
+      }
 
-    if (!functionDef.variables || !functionDef.expression) {
-      throw new Error('Bad mapping for: ' + this.value_);
-    }
-    if (functionDef.variables.length !== this.children_.length) {
-      throw new Error('Bad mapping for: ' + this.value_);
+      // We're calling a new function, so it gets a new local scope.
+      var newLocalMapping = {};
+      functionDef.variables.forEach(function (variable, index) {
+        var childVal = this.getChildValue(index);
+        newLocalMapping[variable] = utils.valueOr(localMapping[childVal], childVal);
+      }, this);
+      return functionDef.expression.evaluate(globalMapping, newLocalMapping);
     }
 
-    // We're calling a new function, so it gets a new local scope.
-    var newLocalMapping = {};
-    functionDef.variables.forEach(function (variable, index) {
-      var childVal = this.getChildValue(index);
-      newLocalMapping[variable] = utils.valueOr(localMapping[childVal], childVal);
-    }, this);
-    return functionDef.expression.evaluate(gloablMapping, newLocalMapping);
-  }
+    if (type === ValueType.NUMBER) {
+      return { result: this.value_ };
+    }
 
-  if (type === ValueType.NUMBER) {
-    return this.value_;
-  }
+    if (type !== ValueType.ARITHMETIC) {
+      throw new Error('Unexpected');
+    }
 
-  if (type !== ValueType.ARITHMETIC) {
-    throw new Error('Unexpected error');
-  }
+    var left = this.children_[0].evaluate(globalMapping, localMapping);
+    var right = this.children_[1].evaluate(globalMapping, localMapping);
 
-  var left = this.children_[0].evaluate(gloablMapping, localMapping);
-  var right = this.children_[1].evaluate(gloablMapping, localMapping);
-
-  switch (this.value_) {
-    case '+':
-      return left + right;
-    case '-':
-      return left - right;
-    case '*':
-      return left * right;
-    case '/':
-      return left / right;
-    default:
-      throw new Error('Unknown operator: ' + this.value_);
+    if (left.err || right.err) {
+      return { err: left.err || right.err };
+    }
+    left = left.result;
+    right = right.result;
+
+    switch (this.value_) {
+      case '+':
+        return { result: left + right };
+      case '-':
+        return { result: left - right };
+      case '*':
+        return { result: left * right };
+      case '/':
+        if (right === 0) {
+          throw new DivideByZeroError();
+        }
+        return { result: left / right };
+      default:
+        throw new Error('Unknown operator: ' + this.value_);
```

Comment:
It seems like in certain places we `return {err: someError}` and others we do the same by throwing and subsequently returning. Is there a rule for when to do one or the other, or would it make sense to standardize on one way?